src/core/public/http/http_service.test.ts

```diff
@@ -313,6 +313,153 @@ describe('http requests', () => {
   });
 });
 
+describe('interception', () => {
+  const tap = (handler: Function) => (value: any) => {
+    handler();
+    return value;
+  };
+  const { http } = setup();
+
+  beforeEach(() => {
+    fetchMock.get('*', { foo: 'bar' });
+  });
+
+  afterEach(() => {
+    fetchMock.restore();
+    http.removeAllInterceptors();
+  });
+
+  it('should make request and receive response', async () => {
+    http.intercept({
+      args: tap(Function.prototype),
+      then: tap(Function.prototype),
+    });
+
+    const body = await http.fetch('/my/path');
+
+    expect(fetchMock.called()).toBe(true);
+    expect(body).toEqual({ foo: 'bar' });
+  });
+
+  it('should manipulate fetch arguments', async () => {
+    http.intercept({
+      args: ([path, fetchOptions]) => [
+        path.replace('path', 'route'),
+        {
+          ...fetchOptions,
+          headers: {
+            ...fetchOptions.headers,
+            'Content-Type': 'CustomContentType',
+          },
+        },
+      ],
+    });
+
+    const body = await http.fetch('/my/path');
+
+    expect(fetchMock.called()).toBe(true);
+    expect(body).toEqual({ foo: 'bar' });
+    expect(fetchMock.lastOptions()!.headers).toMatchObject({
+      'Content-Type': 'CustomContentType',
+    });
+    expect(fetchMock.lastUrl()).toBe('http://localhost/myBase/my/route');
+  });
+
+  it('should call interceptors in correct order', async () => {
+    const order: string[] = [];
+
+    http.intercept({
```

Comment:
Could you elaborate on what you mean by this?

Reply:
end user cannot add endpoint-specific interceptor?

Reply:
I will see about passing an options parameter to allow filtering from these methods.

Reply:
If we do want to do non-global interceptors, I think an interesting pattern would be to make them http client specific. Each plugin would get their own instance of the http client that (1) is scoped to their plugin's endpoints; (2) can register its own interceptors that don't affect other client instances. We would still need to have the ability to register global interceptors of course.

(1) would help us with our goal to only expose programmatic APIs between plugins by limiting plugins to only be able to call their own APIs endpoints.

Reply:
It's worth noting, this does *not* need to be implemented in this PR. That said, I think it's worth doing sooner than later.

Reply:
This can now be done by doing an early-return based on request/response information in the interception. We can follow up in the future for plugin-based interception.